### DISCUSSION TITLE
Weird behaviour - mock hangs when being called twice

### DISCUSSION THREAD

**Discussion by dandraka** [pester/Pester#2569] [Q&A] [answered]
Hi all

(I've posted the same question in [Powershell forums](https://forums.powershell.org/t/weird-bahaviour-mock-hangs-when-being-called-twice/24890) but on a second thought, here is probably the best place to ask)

I’m assigned to write a bunch of tests on an existing codebase and a number of my tests hang.

The code, simplified to show the issue, is as follows:

File MySharepointModule.ps1 :

```
function Update-SharepointList {
  # do some stuff
  Set-PnPListItem -someparams
}
```
File MyScript.ps1 :

```
Import-Module -Name  '.\Modules\MySharepointModule.ps1'
# do lots of stuff
Set-PnPListItem -someparams

# do lots of stuff
Update-SharepointList
```

In order to avoid setting up a Sharepoint site, I’m mocking Set-PnPListItem.

```
Mock Set-PnPListItem { 
    # store args into a variable
}
```

It works for the first call. But on the second call, where it’s being called from the module, Set-PnPListItem hangs. It doesn’t even throw an error -just hangs. And if I need to run the tests again, I have to kill the terminal or restart VSCode.

I’ve tried different variations, even leaving the mock empty. Nothing works, it always hangs. The only way I can get it to work is if I mock Update-SharepointList itself, but that means leaving this code untested.

Is this a known issue, behaviour or bug? Are there any workarounds?

Many thanks in advance

Jim

**Comment by fflaten** [comment, marked as answer]
Hi. This is by design. The mock isn't called when invoked through `Update-SharepointList` as it invokes from another state (the module).

You need two mocks for you sample. 

- The one you have to catch the call done directly in the test.
- Another one specifically for calls made from your module functions

Try adding:

```powershell
Mock Set-PnPListItem -ModuleName MySharepointModule { 
    # store args into a variable
}
```

See https://pester.dev/docs/usage/modules

**Reply by dandraka** [reply]
You're a lifesaver! 🙏🏼
